tsl/src/remote/healthcheck.c

```diff
@@ -0,0 +1,304 @@
+ * status will be "true" while the node is in recovery.
+ *
+ * 4. An optional error message explaining the reason for "healthy" being
+ * false.
+ */
+static HeapTuple
+create_health_check_tuple(const char *data_node, bool in_recovery, TupleDesc tupdesc)
+{
+	Datum values[Natts_health];
+	bool nulls[Natts_health];
+
+	memset(nulls, true, sizeof(nulls));
+
+	if (NULL != data_node)
+	{
+		NameData *node_name = palloc(sizeof(NameData));
+		namestrcpy(node_name, data_node);
+		values[AttrNumberGetAttrOffset(Anum_health_node_name)] = NameGetDatum(node_name);
+		nulls[AttrNumberGetAttrOffset(Anum_health_node_name)] = false;
+	}
+
+	values[AttrNumberGetAttrOffset(Anum_health_healthy)] = BoolGetDatum(true);
+	nulls[AttrNumberGetAttrOffset(Anum_health_healthy)] = false;
+	values[AttrNumberGetAttrOffset(Anum_health_in_recovery)] = BoolGetDatum(in_recovery);
+	nulls[AttrNumberGetAttrOffset(Anum_health_in_recovery)] = false;
+
+	return heap_form_tuple(tupdesc, values, nulls);
+}
+
+static IOFuncSelector
+get_io_func_selector_from_format(int format)
+{
+	switch (format)
+	{
+		case 0: /* text format */
+			return IOFunc_input;
+		case 1: /* binary format */
+			return IOFunc_receive;
+		default:
+			/* reserved for future, so likely won't happen */
+			ereport(ERROR,
+					(errcode(ERRCODE_INTERNAL_ERROR),
+					 errmsg("unexpected format of data node response")));
+	}
+
+	pg_unreachable();
+	/* Silence Windows compiler */
+	return IOFunc_input;
+}
+
+static void
+fill_in_result_error(Datum values[Natts_health], bool nulls[Natts_health], const char *errormsg)
+{
+	values[AttrNumberGetAttrOffset(Anum_health_error)] = CStringGetTextDatum(errormsg);
+	nulls[AttrNumberGetAttrOffset(Anum_health_error)] = false;
+}
+
+Datum
+ts_dist_health_check(PG_FUNCTION_ARGS)
+{
+	DistCmdResult *result;
+	FuncCallContext *funcctx;
+	HeapTuple tuple = NULL;
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		MemoryContext oldcontext;
+		TupleDesc tupdesc;
+
+		funcctx = SRF_FIRSTCALL_INIT();
+
+		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+
+		if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("function returning record called in context "
+							"that cannot accept type record")));
+
+		funcctx->tuple_desc = BlessTupleDesc(tupdesc);
+
+		if (dist_util_membership() == DIST_MEMBER_ACCESS_NODE)
+		{
+			List *data_node_list;
+			const char *search_path;
+			StringInfo cmd = makeStringInfo();
+			Oid fnamespaceid = get_func_namespace(fcinfo->flinfo->fn_oid);
+
+			appendStringInfo(cmd,
+							 "SELECT * FROM %s.%s()",
+							 get_namespace_name(fnamespaceid),
+							 get_func_name(fcinfo->flinfo->fn_oid));
+			data_node_list = data_node_get_node_name_list();
+			search_path = GetConfigOption("search_path", false, false);
+			result = ts_dist_cmd_invoke_on_data_nodes_using_search_path(cmd->data,
```

Comment:
not sure we need to set search_path here

Reply:
Removed.